compiler/IREmitter/Payload/codegen-payload.c

```diff
@@ -1277,6 +1281,80 @@ VALUE sorbet_rb_array_empty(VALUE recv, ID fun, int argc, const VALUE *const res
     return Qfalse;
 }
 
+// This is the no-block version of rb_ary_uniq: https://github.com/ruby/ruby/blob/ruby_2_7/array.c#L5018-L5041
+SORBET_INLINE
+VALUE sorbet_rb_array_uniq(VALUE recv, ID fun, int argc, const VALUE *const restrict argv, BlockFFIType blk,
+                           VALUE closure) {
+    rb_check_arity(argc, 0, 0);
+    VALUE ary = recv;
+
+    VALUE hash, uniq;
+
+    if (RARRAY_LEN(ary) <= 1) {
+        hash = 0;
+        uniq = rb_ary_dup(ary);
+    }
+    else {
+        hash = sorbet_ary_make_hash(ary);
+        uniq = rb_hash_values(hash);
+    }
+    sorbet_rbasic_set_class(uniq, rb_obj_class(ary));
+    if (hash) {
+        sorbet_ary_recycle_hash(hash);
+    }
+
+    return uniq;
+}
+
+// This is the block version of rb_ary_uniq: https://github.com/ruby/ruby/blob/ruby_2_7/array.c#L5018-L5041
+SORBET_INLINE
+VALUE sorbet_rb_array_uniq_withBlock(VALUE recv, ID fun, int argc, const VALUE *const restrict argv, BlockFFIType blk,
+                                     const struct rb_captured_block *captured, VALUE closure, int numPositionalArgs) {
+    rb_check_arity(argc, 0, 0);
+    VALUE ary = recv;
+
+    // must push a frame for the captured block
+    sorbet_pushBlockFrame(captured);
+
+    VALUE hash, uniq;
+
+    if (RARRAY_LEN(ary) <= 1) {
+        hash = 0;
+        uniq = rb_ary_dup(ary);
+    }
+    else {
+        // inline of ary_tmp_hash_new
+        long size = RARRAY_LEN(ary);
+        hash = rb_hash_new_with_size(size);
+
+        RBASIC_CLEAR_CLASS(hash);
+
+        // inline of ary_make_hash_by
+        for (int i = 0; i < RARRAY_LEN(ary); ++i) {
+            // inline of rb_ary_elt
+            VALUE v;
+            long len = RARRAY_LEN(ary);
+            if (len == 0) v = Qnil;
+            else if (i < 0 || len <= i) {
+                v = Qnil;
+            }
+            else v = RARRAY_AREF(ary, i);
```

Comment:
Would you mind putting curly braces around these conditionals? I know the ruby vm has a lot of uncurlied blocks, but I prefer to avoid them when possible :)

I also wonder if the check for `RARRAY_LEN` on line 1336 is necessary. I know it's inherited from inlining `rb_ary_elt`, but wouldn't the loop body not execute if it was 0?

Reply:
Oh you're right about the `RARRAY_LEN` thing too. For a moment I was scratching my head because I wasn't sure if mutating the array inside the block could cause `i` to be out of bounds, but we're re-evaluating `RARRAY_LEN` in the conditional right before we jump in here, so that doesn't appear to be possible. Will simplify this.